### ISSUE TITLE
Allow using different service accounts on GCE

### ISSUE THREAD

**Issue by jdoliner** [pachyderm/pachyderm#2538]
It should be as easy as changing this line:

```
cloud.WithTokenSource(google.ComputeTokenSource(""))
```

to not use an `""`.

**Comment by shebiki** [comment]
I'm a little confused by the various access methods but it looks to me like access scopes are now deprecated[1] and instead Pachyderm should encourage users to create a GCE service account and accept a configuration[2] of the `credentials.json` for that service account that is uses on any pods that need write access to Google Cloud Storage.

1: https://cloud.google.com/compute/docs/access/service-accounts#accesscopesiam
2: https://godoc.org/cloud.google.com/go#example-package--ServiceAccountFile

**Comment by atombender** [comment]
Also, the official Google recommendation (see [Securing containers with Kubernetes Engine 1.8](
https://cloudplatform.googleblog.com/2017/11/precious-cargo-securing-containers-with-Kubernetes-Engine-18.html)) is to reduce the default permissions given to individual nodes:

> The principle of least privilege helps to reduce the "blast radius" of a potential compromise, by granting each component only the minimum permissions required to perform its function. Should one component become compromised, least privilege makes it much more difficult to chain attacks together and escalate permissions.
>
> Each Kubernetes Engine node has a Service Account associated with it. You’ll see the Service Account user listed in the IAM section of the Cloud Console as “Compute Engine default service account.” This account has broad access by default, making it useful to wide variety of applications, but has more permissions than you need to run your Kubernetes Engine cluster.
>
> **We recommend you create and use a minimally privileged service account to run your Kubernetes Engine Cluster instead of the Compute Engine default service account**.

(Emphasis mine.)

**Comment by atombender** [comment]
As far as I can see, at least with the current (0.17) version of the Google SDK, you should instantiate `client.NewClient` without options. The current call is:

```go
	client, err := storage.NewClient(
		ctx,
		cloud.WithTokenSource(google.ComputeTokenSource("")),
		cloud.WithScopes(storage.ScopeFullControl),
	)
```

But reading the code, it looks like the absence of a token source will cause the client to call `google.FindDefaultCredentials`, which goes through the standard Google SDK ways of finding credentials:

```
//   1. A JSON file whose path is specified by the
//      GOOGLE_APPLICATION_CREDENTIALS environment variable.
//   2. A JSON file in a location known to the gcloud command-line tool.
//      On Windows, this is %APPDATA%/gcloud/application_default_credentials.json.
//      On other systems, $HOME/.config/gcloud/application_default_credentials.json.
//   3. On Google App Engine it uses the appengine.AccessToken function.
//   4. On Google Compute Engine and Google App Engine Managed VMs, it fetches
//      credentials from the metadata server.
//      (In this final case any provided scopes are ignored.)
```

As for `WithScopes`, the default scope is actually `ScopeFullControl`, so it can be dropped.

In other words, Pachyderm can just use `google.DefaultClient(ctx)` instead of what it does now. On GCE, this would use the metadata API to get the node's credentials (today's behaviour), but also allow Pachyderm to be deployed with other credentials (either OAuth 2.0 token or a service account, as [described here](https://developers.google.com/identity/protocols/application-default-credentials)).

**Comment by shebiki** [comment]
This ticket is related to https://github.com/pachyderm/pachyderm/issues/2578

**Comment by jdoliner** [comment]
Just an update on this, this is being actively worked on and should be available in 1.7.
